draft-ounsworth-cfrg-kem-combiners.mkd

```diff
@@ -228,40 +251,55 @@ ss = kemCombiner(ss_1, ss_2, ..., ss_n)
 
 This document assumes that shared secrets are the output of a KEM, but without loss of generality they MAY also be any other source of cryptographic key material, such as pre-shared keys (PSKs), with PQ/PSK being a quantum-safe migration strategy being made available by some protocols, see for example IKEv2 in [RFC8784].
 
-In general it is desirable to use a multi-PRF as a KEM combiner, meaning a PRF when keyed by any of its single inputs.
+In general it is desirable to use a multi-PRF as a KEM combiner, meaning that the combiner has the properties of a PRF when keyed by any of its single inputs.
 The following simple yet generic construction can be used in all IETF protocols that need to combine the output of two or more KEMs:
 
 ~~~
 KDF(counter || k_1 || ... || k_n || fixedInfo, outputBits)
 ~~~
-{: title="general KEM combiner construction" }
+{: #tab-kemCombiner title="general KEM combiner construction" }
 
 where:
 
 - `KDF` represents a suitable choice of cryptographic key derivation function,
-- `k_i` represent the constant-length input keys,
+- `k_i` represent the constant-length input keys and is discussed in {{sec-k_i-construction}},
 - `fixedInfo` is some protocol-specific KDF binding,
 - `counter` parameter is instantiation-specific and is discussed in {{sec-instantiation}}.
-- `outputBits` determines the length of the key,
+- `outputBits` determines the length of the key to output,
 - `||` represents concatenation.
 
-In section {{sec-instantiation}} are listed several possible practical instantiations, in compliance with NIST SP-800 56Cr2 {{SP800-56C}}.
-The shared secret `ss` MAY be used as a Key Encryption Key (KEK) for key encapsulation.
+In {{sec-instantiation}} are listed several possible practical instantiations, in compliance with NIST SP-800 56Cr2 {{SP800-56C}}.
+The shared secret `ss` MAY be used directly as a symmetric key, for example as a MAC key or as a Key Encryption Key (KEK) for key encapsulation.
+
+## k_i construction {#sec-k_i-construction}
 
-Each `k_i` MUST be constant in length, therefore the secret shares `ss_i` can be used directly only if they are guaranteed to be constant length. For all other cases, it is REQUIRED to hash them first:
+Following the guidance of Giacon et al. [KEMCombiners], we wish for a KEM combiner that is (CCA) secure as long as at least one of the ingredient KEMs is. Ir order to protect against chosen ciphertext attacks, it is necessary to include both the shared secret `ss_i` and its corresponding ciphertext `ct_i`. In addition, each `k_i` MUST be constant in length to prevent abuse of the underlying hash function, therefore the secret shares `ss_i || ct_i` MUST be hashed prior to inclusion in the overall construction described in {{tab-kemCombiner}}:
```

Comment:
```suggestion
Following the guidance of Giacon et al. [KEMCombiners], we wish for a KEM combiner that is CCA-secure as long as at least one of the ingredient KEMs is. In order to protect against chosen ciphertext attacks, it is necessary to include both the shared secret `ss_i` and its corresponding ciphertext `ct_i` in the `k_i`. In addition, each `k_i` MUST be constant in length to prevent abuse of the underlying hash function, therefore the secret shares `ss_i` and ciphertexts `ct_i` MUST be hashed prior to inclusion in the overall construction described in {{tab-kemCombiner}}:
```
Giacon et al focus on CCA, so I removed the brackets here. Is it really "CCA-secure", or rather "CCA secure". I opted for the first option (would need alignment to the abstract).

I am in favour of hashing as a must, since even if you are opened to constant length considerations an implementer has to implement length checks if he wants to have a secure implementation.

I am a bit concerned with the combination of not hashing ss_i || ct_i and using SHAKE, since SHAKE is an XOF and no approved hash function, so additional security considerations are necessary. I'll take up the discussion where SHAKE is listed.

Reply:
So, I want SHAKE just so I can get 384-bit outputs. I could equally do SHA3-512 truncated to 384 if that is more academically-sound.

Reply:
You can do KMAC with output size 384 bits. KMAC is internally just SHAKE only with some further internal encodings and paddings, that e.g. break the truncation property.

Let's just keep SHAKE for the moment, I find it an interesting option and we should give it some thought.